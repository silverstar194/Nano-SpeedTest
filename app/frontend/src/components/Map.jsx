import React from 'react';
import { compose, withProps } from 'recompose';
import { withGoogleMap, GoogleMap, Polyline, Marker } from 'react-google-maps';
import '../styles/GoogleMaps.css';

class TestMap extends React.Component {
    state = {
        mapLoaded: false
    };

    // Updates the map window once it is loaded to only show the distance between the two cities
    recomputeBounds() {
        const {cities} = this.props;
        if (!Object.keys(cities).length) return;

        const bounds = new window.google.maps.LatLngBounds();
        Object.values(cities).map((city) => bounds.extend(city.coords));
        this.refs.map.fitBounds(bounds, 3);
        this.setState({mapLoaded: true});
    }

    render() {
        const {cities} = this.props;
        const {mapLoaded} = this.state;
        return (
            <GoogleMap
                ref='map'
                defaultZoom={2}
                defaultCenter={{ lat: 0, lng: 0 }}
                defaultOptions={{ // disable moving and zooming the map
                    zoomControl: false,
                    gestureHandling: 'none',
                    draggableCursor: 'default',
                    disableDefaultUI: true
                }}
                onTilesLoaded={() => this.recomputeBounds()}
            >
            { // the map needs to render before recomputing and drawing markers and lines
                (mapLoaded && Object.keys(cities).length) ? (
                    <div>
                        <Polyline
                            path={[cities.origin.coords, cities.destination.coords]}
                            options={{
                                clickable: false,
                                strokeOpacity: 1,
                                strokeWeight: 2,
                            }}
                        />
                        {Object.values(cities).map((city, index) =>
                            <Marker
                                key={city.name || index}
                                title={city.name}
                                position={city.coords}
                                clickable={false}
                            />
                        )}
                    </div>
                ) : null

            }

            </GoogleMap>
        );
  }
}

const Map = compose(
    withProps({
        loadingElement: <div className={'GoogleMapsLoading'} />,
        containerElement: <div className={'GoogleMapsContainer'} />,
        mapElement: <div className={'GoogleMapsElement'} />
    }),
    withGoogleMap
)(TestMap);

export default Map;
