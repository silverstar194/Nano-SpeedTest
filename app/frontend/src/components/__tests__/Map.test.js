import React from 'react';
import renderer from 'react-test-renderer';
import Map from '../Map';
import { shallow } from 'enzyme';

it('shallow renders without crashing', () => {
    shallow(<Map />);
});

it(`renders snapshot correctly`, () => {
    const tree = renderer.create(
        <Map />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
