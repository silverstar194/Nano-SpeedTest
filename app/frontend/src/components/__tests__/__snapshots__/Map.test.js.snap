// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `
<div
  className="GoogleMapsContainer"
>
  <div
    className="GoogleMapsElement"
  />
  <div />
</div>
`;
